kernel/src/syscall/stat.rs

```diff
@@ -135,10 +193,151 @@ impl From<Metadata> for Stat {
+            stx_ctime: StatxTimestamp::from(info.ctime),
+            stx_mtime: StatxTimestamp::from(info.ctime),
+            stx_rdev_major: rdevid.major(),
+            stx_rdev_minor: rdevid.minor(),
+            stx_dev_major: devid.major(),
+            stx_dev_minor: devid.minor(),
+            stx_mnt_id: 0,
+            stx_dio_mem_align: 0,
+            stx_dio_offset_align: 0,
+            __spare3: [0; 12],
+        }
+    }
+}
+
+/// Statx Timestamp (seconds and nanoseconds)
+#[derive(Debug, Clone, Copy, Pod, Default)]
+#[repr(C)]
+pub struct StatxTimestamp {
+    /// Seconds
+    tv_sec: i64,
+    /// Nanoseconds
+    tv_nsec: u32,
+    __reserved: i32,
+}
+
+impl From<Duration> for StatxTimestamp {
+    fn from(duration: Duration) -> Self {
+        Self {
+            tv_sec: duration.as_secs() as i64,
+            tv_nsec: duration.subsec_nanos() as u32,
+            __reserved: 0,
+        }
+    }
+}
+
+bitflags! {
     struct StatFlags: u32 {
         const AT_EMPTY_PATH = 1 << 12;
         const AT_NO_AUTOMOUNT = 1 << 11;
         const AT_SYMLINK_NOFOLLOW = 1 << 8;
     }
 }
+
+bitflags! {
+    struct StatxFlags: u32 {
+        const AT_EMPTY_PATH = 1 << 12;
+        const AT_NO_AUTOMOUNT = 1 << 11;
+        const AT_SYMLINK_NOFOLLOW = 1 << 8;
+        const AT_STATX_SYNC_AS_STAT = 0;
+        const AT_STATX_FORCE_SYNC = 1 << 13;
+        const AT_STATX_DONT_SYNC = 1 << 14;
+    }
+}
+
+bitflags! {
+    pub struct StatxMask: u32 {
+        const STATX_TYPE            = 0x00000001;  // Want stx_mode & S_IFMT
+        const STATX_MODE            = 0x00000002;  // Want stx_mode & ~S_IFMT
+        const STATX_NLINK           = 0x00000004;  // Want stx_nlink
+        const STATX_UID             = 0x00000008;  // Want stx_uid
+        const STATX_GID             = 0x00000010;  // Want stx_gid
+        const STATX_ATIME           = 0x00000020;  // Want stx_atime
+        const STATX_MTIME           = 0x00000040;  // Want stx_mtime
+        const STATX_CTIME           = 0x00000080;  // Want stx_ctime
+        const STATX_INO             = 0x00000100;  // Want stx_ino
+        const STATX_SIZE            = 0x00000200;  // Want stx_size
+        const STATX_BLOCKS          = 0x00000400;  // Want stx_blocks
+        const STATX_BASIC_STATS     = 0x000007ff;  // All of the above (stx_mode, stx_nlink, etc.)
+        const STATX_BTIME           = 0x00000800;  // Want stx_btime
+        const STATX_ALL             = 0x00000fff;  // Deprecated: The same as STATX_BASIC_STATS | STATX_BTIME
+        const STATX_MNT_ID          = 0x00001000;  // Want stx_mnt_id
+        const STATX_DIOALIGN        = 0x00002000;  // Want stx_dio_mem_align and stx_dio_offset_align
+        const STATX_MNT_ID_UNIQUE   = 0x00004000;  // Want unique stx_mnt_id
+        const STATX_RESERVED		= 0x80000000;  // Reserved for future struct statx expansion
+        const STATX_CHANGE_COOKIE   = 0x40000000;  // Want/got stx_change_attr
+    }
```

Comment:
Reading up to this point, it seems that the `statx` syscall is quite independent of `stat`, with its own structs, flags, and masks. Perhaps we could place the logic in a new file named `statx.rs`.